// node-easywechat/dist/OpenPlatform/Authorizer/MiniApp/Utils.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const Utils_1 = __importDefault(require("../../../MiniApp/Utils"));
class Utils extends Utils_1.default {
    constructor(app) {
        super(app);
        this.app = null;
        this.app = app;
    }
    /**
     * 代理小程序登录
     * @see https://developers.weixin.qq.com/doc/oplatform/openApi/OpenApiDoc/miniprogram-management/login/thirdpartyCode2Session.html
     * @param code
     * @returns
     */
    code2Session(code) {
        return __awaiter(this, void 0, void 0, function* () {
            let client = this.app.getHttpClient();
            let response = yield client.request('GET', '/sns/component/jscode2session', {
                params: {
                    component_access_token: yield this.app.getComponentApp().getAccessToken().getToken(),
                    component_appid: this.app.getComponentApp().getAccount().getAppId(),
                    appid: this.app.getAccount().getAppId(),
                    js_code: code,
                    grant_type: 'authorization_code',
                }
            });
            let data = response.toObject();
            if (!data['openid']) {
                throw new Error(`code2Session error: ${JSON.stringify(data)}`);
            }
            return data;
        });
    }
}
;
module.exports = Utils;
