// node-easywechat/dist/OpenPlatform/AuthorizerAccessToken.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
class AuthorizerAccessToken {
    constructor(appId, accessToken) {
        this.appId = appId;
        this.accessToken = accessToken;
    }
    /**
     * 获取appId
     * @returns
     */
    getAppId() {
        return this.appId;
    }
    getToken() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.accessToken;
        });
    }
    toQuery() {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                'access_token': yield this.getToken(),
            };
        });
    }
    /**
     * 转为字符串
     * @returns
     */
    toString() {
        return this.accessToken;
    }
}
module.exports = AuthorizerAccessToken;
