// node-easywechat/dist/OpenWork/AuthorizerAccessToken.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const HttpClient_1 = __importDefault(require("../Core/HttpClient/HttpClient"));
class AuthorizerAccessToken {
    constructor(corpId, permanentCodeOrAccessToken, suiteAccessToken = null, key = null, cache = null, httpClient = null) {
        this.corpId = corpId;
        this.permanentCodeOrAccessToken = permanentCodeOrAccessToken;
        this.suiteAccessToken = suiteAccessToken;
        this.key = key;
        this.cache = cache;
        this.httpClient = httpClient;
        if (!this.httpClient) {
            this.httpClient = HttpClient_1.default.create({
                baseURL: 'https://qyapi.weixin.qq.com/',
            });
        }
    }
    /**
     * 获取appId
     * @returns
     */
    getCorpId() {
        return this.corpId;
    }
    /**
     * 获取access_token的缓存名称
     * @returns
     */
    getKey() {
        if (!this.key) {
            this.key = `open_work.authorizer.access_token.${this.corpId}.${this.permanentCodeOrAccessToken}`;
        }
        return this.key;
    }
    /**
     * 设置access_token的缓存名称
     * @param key
     * @returns
     */
    setKey(key) {
        this.key = key;
        return this;
    }
    getToken() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.suiteAccessToken) {
                return this.permanentCodeOrAccessToken;
            }
            let token = '';
            if (this.cache) {
                token = yield this.cache.get(this.getKey());
            }
            if (!!token && typeof token === 'string') {
                return token;
            }
            return this.refresh();
        });
    }
    toQuery() {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                access_token: yield this.getToken(),
            };
        });
    }
    refresh() {
        return __awaiter(this, void 0, void 0, function* () {
            let response = (yield this.httpClient.request('post', 'cgi-bin/service/get_corp_token', {
                params: {
                    suite_access_token: yield this.suiteAccessToken.getToken(),
                },
                json: {
                    auth_corpid: this.corpId,
                    permanent_code: this.permanentCodeOrAccessToken,
                }
            })).toObject();
            if (!response['access_token']) {
                throw new Error('Failed to get access_token: ' + JSON.stringify(response));
            }
            if (this.cache) {
                yield this.cache.set(this.getKey(), response['access_token'], parseInt(response['expires_in']) - 100);
            }
            return response['access_token'];
        });
    }
    /**
     * 转为字符串
     * @returns
     */
    toString() {
        return this.getToken();
    }
}
module.exports = AuthorizerAccessToken;
